### PR TITLE
[generator] Incremental updates via bsdiff.

### PR DESCRIPTION
**Pull request by mpimenov**
https://jira.mail.ru/browse/MAPSME-5098
https://jira.mail.ru/browse/MAPSME-5325

### CODE REVIEW

File: 3party/bsdiff-courgette/bsdiff/bsdiff.h [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+
+  my::Timer bsdiff_timer;
+
+  std::array<MemStream, kNumStreams> mem_streams;
+  auto & control_stream_copy_counts = mem_streams[0];
+  auto & control_stream_extra_counts = mem_streams[1];
+  auto & control_stream_seeks = mem_streams[2];
+  auto & diff_skips = mem_streams[3];
+  auto & diff_bytes = mem_streams[4];
+  auto & extra_bytes = mem_streams[5];
```
Comment (ygorshenin) [inline 3party/bsdiff-courgette/bsdiff/bsdiff.h]: ASSERT that kNumStreams is at least 6.

File: 3party/bsdiff-courgette/bsdiff/bsdiff.h [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+
+  const int old_size = static_cast<int>(old_source.Size());
```
Comment (ygorshenin) [inline 3party/bsdiff-courgette/bsdiff/bsdiff.h]: Why not size_t?

Comment (ygorshenin) [inline 3party/bsdiff-courgette/bsdiff/bsdiff.h]: Okay, existing code.

File: generator/mwm_diff/diff.cpp [hunk 3]
```diff
@@ -2,53 +2,121 @@
...
+    FileReader diffFileReader(diffPath);
 
-bool ApplyDiff(string const & /* oldMwmPath */, string const & newMwmPath, string const & diffPath)
-{
-  return CopyFile(diffPath, newMwmPath);
+    ReaderSource<FileReader> diffFileSource(diffFileReader);
+    auto const header = ReadPrimitiveFromSource<uint32_t>(diffFileSource);
+    if (header != kLatestVersion)
+    {
+      LOG(LERROR, ("Unknown version format of mwm diff:", header));
```
Comment (ygorshenin) [inline generator/mwm_diff/diff.cpp]: We may try to apply not only latest-version diffs. In general, your code should always generate diff in the latest format, but must be able to apply diffs of all formats.

File: 3party/bsdiff-courgette/bsdiff/bsdiff.h (hunk 1 above)
Comment (mpimenov) [inline 3party/bsdiff-courgette/bsdiff/bsdiff.h]: done

File: generator/mwm_diff/diff.cpp (hunk 3 above)
Comment (mpimenov) [inline generator/mwm_diff/diff.cpp]: done

File: 3party/bsdiff-courgette/bsdiff/bsdiff.h [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+      uint32_t extra_count = gap;
+      int32_t seek_adjustment = ((match.pos - lenb) - (lastpos + lenf));
+
+      WriteVarUint(control_stream_copy_counts.GetWriter(), copy_count);
+      WriteVarUint(control_stream_extra_counts.GetWriter(), extra_count);
+      WriteVarInt(control_stream_seeks.GetWriter(), seek_adjustment);
+
+      ++control_length;
+
+#ifdef DEBUG_bsmedberg
```
Comment (mpimenov) [inline 3party/bsdiff-courgette/bsdiff/bsdiff.h]: Existing code. Looks useful enough to leave it as is.

File: 3party/bsdiff-courgette/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,22 @@
+project(bsdiff)
+
+set(CMAKE_PREFIX_PATH ./)
```
Comment (mpimenov) [inline 3party/bsdiff-courgette/CMakeLists.txt]: Removed.

File: generator/mwm_diff/diff.cpp (hunk 3 above)
Comment (mpimenov) [inline generator/mwm_diff/diff.cpp]: done

File: 3party/bsdiff-courgette/bsdiff/bsdiff.h [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+//                --Stephen Adams <sra@chromium.org>
+// 2010-05-26 - Use a paged array for V and I. The address space may be too
+//              fragmented for these big arrays to be contiguous.
+//                --Stephen Adams <sra@chromium.org>
+// 2015-08-03 - Extract qsufsort portion to a separate file.
+//                --Samuel Huang <huangs@chromium.org>
+// 2015-08-12 - Interface change to search().
+//                --Samuel Huang <huangs@chromium.org>
+// 2016-07-29 - Replacing qsufsort with divsufsort.
+//                --Samuel Huang <huangs@chromium.org>
```
Comment (mpimenov) [inline 3party/bsdiff-courgette/bsdiff/bsdiff.h]: Appended to the changelog.
I'm not sure on how we should change the copyright message (and whether we should change it at all).

**Comment by mpimenov** [comment]
@mgsergio @ygorshenin PTAL

File: generator/mwm_diff/diff.cpp [hunk 7]
```diff
@@ -2,34 +2,71 @@
...
+#include "3party/bsdiff-courgette/bsdiff/bsdiff.h"
+
 using namespace std;
 
 namespace
 {
-// We could use my::CopyFileX but it fails if the file at |srcFile| is empty.
-bool CopyFile(string const & srcFile, string const & dstFile)
+// Format Version 0: bsdiff+gzip.
+uint32_t const kLatestVersion = 0;
```
Comment (ygorshenin) [inline generator/mwm_diff/diff.cpp]:
```c++
enum Version
{
  VERSION_V0,
  VERSION_LATEST = VERSION_V0
};
```
?

File: generator/mwm_diff/diff.cpp [hunk 8]
```diff
@@ -2,34 +2,71 @@
...
   }
-  catch (Writer::Exception const & e)
+
+  using Deflate = coding::ZLib::Deflate;
+  Deflate deflate(Deflate::Format::ZLib, Deflate::Level::BestCompression);
+
+  vector<uint8_t> deflatedDiffBuf;
+  deflate(diffBuf.data(), diffBuf.size(), back_inserter(deflatedDiffBuf));
+
+  uint32_t const header = 0;  // Version.
```
Comment (ygorshenin) [inline generator/mwm_diff/diff.cpp]: Please, use enum. Also, it's not a header, it's a version.

File: generator/mwm_diff/diff.cpp (hunk 7 above)
Comment (mpimenov) [inline generator/mwm_diff/diff.cpp]: done

File: generator/mwm_diff/diff.cpp (hunk 8 above)
Comment (mpimenov) [inline generator/mwm_diff/diff.cpp]: done

**Comment by mpimenov** [comment]
@mgsergio @ygorshenin PTAL

**Comment by mpimenov** [comment]
DUDE JTTP

**Comment by mgsergio** [comment]
LGTM

**Pull request merged**